sys/net/network_layer/ng_ndp/ng_ndp.c

```diff
@@ -0,0 +1,950 @@
+        /* TL2A is not supplied and tgt is not anycast */
+        adv_flags |= NG_NDP_NBR_ADV_FLAGS_O;
+    }
+
+    hdr = ng_ndp_nbr_adv_build(adv_flags, tgt, pkt);
+
+    if (hdr == NULL) {
+        DEBUG("ndp: error allocating Neighbor advertisement.\n");
+        ng_pktbuf_release(pkt);
+        return;
+    }
+
+    pkt = hdr;
+    hdr = ng_ipv6_hdr_build(pkt, NULL, 0, (uint8_t *)dst,
+                            sizeof(ng_ipv6_addr_t));
+
+    if (hdr == NULL) {
+        DEBUG("ndp: error allocating IPv6 header.\n");
+        ng_pktbuf_release(pkt);
+        return;
+    }
+
+    ((ng_ipv6_hdr_t *)hdr->data)->hl = 255;
+
+    pkt = hdr;
+    /* add netif header for send interface specification */
+    hdr = ng_netif_hdr_build(NULL, 0, NULL, 0);
+
+    if (hdr == NULL) {
+        DEBUG("ndp: error allocating netif header.\n");
+        return;
+    }
+
+    LL_PREPEND(pkt, hdr);
+
+    ((ng_netif_hdr_t *)hdr->data)->if_pid = iface;
+
+    if (ng_ipv6_netif_addr_is_non_unicast(tgt)) {
+        /* avoid collision for anycast addresses */
+        timex_t delay = { _rand(0, NG_NDP_MAX_AC_TGT_DELAY), 0 };
+        ng_ipv6_nc_t *nc_entry = ng_ipv6_nc_get(iface, tgt);
+
+        /* nc_entry must be set so no need to check it */
+        _send_delayed(&nc_entry->nbr_adv_timer, delay, pkt);
+    }
+    else {
+        ng_netapi_send(ng_ipv6_getpid(), pkt);
+    }
+}
+
+static inline ng_pktsnip_t *_opt_l2a_build(uint8_t type, const uint8_t *l2addr,
+                                           uint8_t l2addr_len, ng_pktsnip_t *next)
+{
+    ng_ndp_opt_t *l2a_opt;
+    ng_pktsnip_t *pkt = ng_ndp_opt_build(type, sizeof(ng_ndp_opt_t) + l2addr_len,
+                                         next);
+
+    if (pkt == NULL) {
+        return NULL;
+    }
+
+    l2a_opt = pkt->data;
+    memset(l2a_opt + 1, 0, pkt->size - sizeof(ng_ndp_opt_t));
+    memcpy(l2a_opt + 1, l2addr, l2addr_len);
+
+    return pkt;
+}
+
+ng_pktsnip_t *ng_ndp_opt_sl2a_build(const uint8_t *l2addr, uint8_t l2addr_len,
+                                    ng_pktsnip_t *next)
+{
+    DEBUG("ndp: building source link-layer address option\n");
+
+    return _opt_l2a_build(NG_NDP_OPT_SL2A, l2addr, l2addr_len, next);
+}
+
+ng_pktsnip_t *ng_ndp_opt_tl2a_build(const uint8_t *l2addr, uint8_t l2addr_len,
+                                    ng_pktsnip_t *next)
+{
+    DEBUG("ndp: building target link-layer address option\n");
+
+    return _opt_l2a_build(NG_NDP_OPT_TL2A, l2addr, l2addr_len, next);
+}
+
+/* internal functions */
+/* packet queue node allocation */
+static ng_pktqueue_node_t *_alloc_pkt_node(ng_pktsnip_t *pkt)
+{
+    for (size_t i = 0; i < sizeof(_pkt_nodes); i++) {
+        if (_pkt_nodes[i].data == NULL) {
+            ng_pktqueue_node_init(_pkt_nodes + i);
+            _pkt_nodes[i].data = pkt;
+
+            return _pkt_nodes + i;
+        }
+    }
+
+    return NULL;
+}
+
+static inline void _free_pkt_node(ng_pktqueue_node_t *node)
+{
+    node->data = NULL;
```

Comment:
It's an inline function, so nothing is lost vs. "what does `node->data = NULL` mean".